paparazzi2/src/main/java/app/cash/paparazzi/Device.kt

```diff
@@ -0,0 +1,556 @@
+
+  val context: Context
+    get() = RenderAction.getCurrentContext()
+
+  private val contentRoot = """
+        |<?xml version="1.0" encoding="utf-8"?>
+        |<${if (hasComposeRuntime) "app.cash.paparazzi.internal.ComposeViewAdapter" else "FrameLayout"}
+        |     xmlns:android="http://schemas.android.com/apk/res/android"
+        |              android:layout_width="${if (frameSpec.renderingMode.horizAction == SessionParams.RenderingMode.SizeAction.SHRINK) "wrap_content" else "match_parent"}"
+        |              android:layout_height="${if (frameSpec.renderingMode.vertAction == SessionParams.RenderingMode.SizeAction.SHRINK) "wrap_content" else "match_parent"}"/>
+  """.trimMargin()
+
+  init {
+    registerFontLookupInterceptionIfResourceCompatDetected()
+    registerViewEditModeInterception()
+    registerMatrixMultiplyInterception()
+    registerChoreographerDelegateInterception()
+    registerServiceManagerInterception()
+    registerIInputMethodManagerInterception()
+
+    ByteBuddyAgent.install()
+    InterceptorRegistrar.registerMethodInterceptors()
+
+    prepare()
+  }
+
+  override fun snapshot(composable: @Composable () -> Unit, timestampMillis: Long): Snapshot {
+    val hostView = ComposeView(context)
+    hostView.setContent(composable)
+
+    return snapshot(hostView, timestampMillis)
+  }
+
+  override fun snapshot(view: View, timestampMillis: Long): Snapshot {
+    val viewGroup = bridgeRenderSession.rootViews[0].viewObject as ViewGroup
+    val modifiedView = prepareSnapshot(view, viewGroup)
+
+    val snapshotTimeNanos = TimeUnit.MILLISECONDS.toNanos(timestampMillis)
+    var snapshotImage: BufferedImage? = null
+    withTime(snapshotTimeNanos) {
+      val result = renderSession.render(true)
+      if (result.status == Result.Status.ERROR_UNKNOWN) {
+        throw result.exception
+      }
+
+      snapshotImage = bridgeRenderSession.image
+      if (validateAccessibility) {
+        require(renderExtensions.isEmpty()) {
+          "Running accessibility validation and render extensions simultaneously is not supported."
+        }
+        validateLayoutAccessibility(modifiedView, snapshotImage)
+      }
+    }
+    viewGroup.removeView(modifiedView)
+    AnimationHandler.sAnimatorHandler.set(null)
+    if (hasComposeRuntime) {
+      forceReleaseComposeReferenceLeaks()
+    }
+    return Snapshot(frameSpec, snapshotImage!!)
+  }
+
+  override fun clip(
+    view: View,
+    clipSpec: ClipSpec
+  ) = Clip(
+    spec = clipSpec,
+    images = flow {
+      val viewGroup = bridgeRenderSession.rootViews[0].viewObject as ViewGroup
+      val modifiedView = prepareSnapshot(view, viewGroup)
+      // Add one to the frame count so we get the last frame. Otherwise a 1 second, 60 FPS animation
+      // our 60th frame will be at time 983 ms, and we want our last frame to be 1,000 ms. This gets
+      // us 61 frames for a 1 second animation, 121 frames for a 2 second animation, etc.
```

Comment:
I like this behaviour! But I’d like to move this math out to the code that produces the ClipSpec?